scripts/generate_can.py

```diff
@@ -1,279 +1,299 @@
+                elif 'scale' in signal and 'offset' in signal:
+                    scale = signal['scale']
+                    offset = signal['offset']
+
+                    # calculate min/max based on scale/offset and length
+                    sig_min = - (2**(length-1)-1) * scale + \
+                        offset if is_signed else offset
+                    sig_max = (2**(length-1)-1) * scale + \
+                        offset if is_signed else (2**length-2) * scale + offset
+
+                # else if only given min/max (no scale/offset)
+                elif 'min' in signal and 'max' in signal:
+                    sig_min = signal['min']
+                    sig_max = signal['max']
+
+                    # calculate scale/offset based on min/max and length
+                    scale = (sig_max - sig_min) / (2**length - 2)
+                    offset = sig_min + (2**(length-1)-1) * \
+                        scale if is_signed else sig_min
+
+                else:
+                    print('Missing fields for signal "' + signal_name +
+                        '" in message "' + message_name + '"!')
+                    print('Supply either signal scale/offset or signal min/max')
+                    print('Script terminated.')
+                    exit()
+
+                # add these missing fields to the yaml
+                signal['scale'] = scale
+                signal['offset'] = offset
+                signal['min'] = sig_min
+                signal['max'] = sig_max
+
+                signals.append(
+                    cantools.database.can.Signal(
+                        name=signal_name,
+                        start=startbit,
+                        length=signal['length'],
+                        byte_order='little_endian',
+                        is_signed=signal['is_signed'],
+                        scale=signal['scale'],
+                        offset=signal['offset'],
+                        minimum=signal['min'],
+                        maximum=signal['max'],
+                        unit=signal['unit'],
+                        choices=signal['values'] if 'values' in signal else None,
+                        comment=signal['comment'],
+                        receivers=message['receivers']
+                    )
+                )
+
+                startbit += signal['length']
+
+            messages.append(
+                cantools.database.can.Message(
+                    frame_id=message['id'],
+                    name=message_name,
+                    length=message['size'],
+                    signals=signals,
+                    comment=message['comment'],
+                    senders=message['senders'],
+                    cycle_time=(None if message['frequency']
+                                == 0 else 1/message['frequency']),
+                )
+            )
+            # change value in "id:" section 
+            message['id'] = CAN_ID(message['id'])

```

Comment:
so does this make it so that ids MUST be specified in hex? and other hex values specified in the yaml aren't converted?

Reply:
This will ensure that the IDs are in hex when the yaml file is re-written. They can be entered in decimal (octo too i think but idk why you'd want to do that), and when the script is run, it will convert everything to hex.

And yes, this will only convert the CAN ID values with a key of "id".

Reply:
I think there may be other fields where this is useful. ex: `canid_filter_mask`